File: src/main/scala/contest/w215/w215_4.scala
package contest.w215

import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

/**
 * [[https://leetcode.com/contest/weekly-contest-215/problems/maximize-grid-happiness/]]
 */
//noinspection DuplicatedCode
class w215_4 extends AnyWordSpec with Matchers {
  /**
   * === 1659. Maximize Grid Happiness ===
   *
   * You are given four integers, `m`, `n`, `introvertsCount`, and `extrovertsCount`.
   * You have an `m x n` grid, and there are two types of people: introverts and extroverts.
   * There are `introvertsCount` introverts and `extrovertsCount` extroverts.
   *
   * You should decide how many people you want to live in the grid and assign each of them one grid cell.
   * Note that you '''do not''' have to have all the people living in the grid.
   *
   * The '''happiness''' of each person is calculated as follows:
   *  - Introverts '''start''' with `120` happiness and '''lose''' `30` happiness for each neighbor (introvert or extrovert).
   *  - Extroverts '''start''' with `40` happiness and '''gain''' `20` happiness for each neighbor (introvert or extrovert).
   *
   * Neighbors live in the directly adjacent cells north, east, south, and west of a person's cell.
   *
   * The '''grid happiness''' is the '''sum''' of each person's happiness.
   * Return ''the '''maximum possible grid happiness'''''.
   *
   * '''Constraints:'''
   *  - `1 <= m, n <= 5`
   *  - `0 <= introvertsCount, extrovertsCount <= min(m * n, 6)`
   */
  object Solution {
    def getMaxGridHappiness(m: Int, n: Int, introvertsCount: Int, extrovertsCount: Int): Int = {
      /**
       * `mem` a ternary flag container of size `n`.
       *
       * `memSize` is a limit on `mem`, we only need to keep track of `n` positions to get upper and left neighbor.
       *
       * ''Trite meaning:'' 0: empty, 1: introvert, 2: extrovert
       *
       * ''Example:''
       * {{{
       * 0d -> 00t
       * 1d -> 01t
       * 2d -> 02t
       * 3d -> 10t
       * 4d -> 11t
       * 5d -> 12t
       * }}}
       */
      val memSize = math.pow(3, n).toInt

      def addToMem(mem: Int, t: Int): Int = (mem * 3 + t) % memSize
      def getLeft(mem: Int): Int = mem % 3
      def getUp(mem: Int): Int = mem / (memSize / 3)

      def intH(n: Int): Int = if (n == 0) 0 else if (n == 1) -60 else -10
      def extH(n: Int): Int = if (n == 0) 0 else if (n == 1) -10 else 40

      val cache = Array.ofDim[Int](m, n, introvertsCount + 1, extrovertsCount + 1, memSize)

      def getMaxHappiness(row: Int, col: Int, is: Int, es: Int, mem: Int): Int = {
        if (is == 0 && es == 0) 0
        else if (col == n) getMaxHappiness(row + 1, 0, is, es, mem)
        else if (row == m) 0
        else if (cache(row)(col)(is)(es)(mem) != 0) cache(row)(col)(is)(es)(mem)
        else {
          val nCol = col + 1
          val up = if (row > 0) getUp(mem) else 0
          val left = if (col > 0) getLeft(mem) else 0

          val r0 =
            getMaxHappiness(row, nCol, is, es, addToMem(mem, 0))
          val r1 = if (is == 0) 0 else {
            val localHappiness = 120 + intH(left) + intH(up)
            getMaxHappiness(row, nCol, is - 1, es, addToMem(mem, 1)) + localHappiness
          }
          val r2 = if (es == 0) 0 else {
            val localHappiness = 40 + extH(left) + extH(up)
            getMaxHappiness(row, nCol, is, es - 1, addToMem(mem, 2)) + localHappiness
          }
          val res = r0 max r1 max r2
          cache(row)(col)(is)(es)(mem) = res
          res
        }
      }

      getMaxHappiness(0, 0, introvertsCount, extrovertsCount, 0)
    }
  }

  object Solution1 {
    def getMaxGridHappiness(m: Int, n: Int, introvertsCount: Int, extrovertsCount: Int): Int = {
      /**
       * `mem` is a sequence of last `n` trites, each takes 2 bits.
       *
       * `memSize` is a limit on `mem`, we only need to keep track of `n` positions to get upper and left neighbor
       *
       * ''Trite meaning:'' 0: empty, 1: introvert, 2: extrovert
       *
       * ''Example:''
       * {{{
       * 0000b -> 00t
       * 0001b -> 01t
       * 0010b -> 02t
       * 0011b -> error
       * 0100b -> 10t
       * 0101b -> 11t
       * }}}
       */
      val memSize = (1 << (n << 1)) - 1

      def addToMem(mem: Int, t: Int): Int = ((mem << 2) | t) & memSize
      def getLeft(mem: Int): Int = mem & 3
      def getUp(mem: Int): Int = mem >> ((n - 1) << 1)

      def intH(n: Int): Int = if (n == 0) 0 else if (n == 1) -60 else -10
      def extH(n: Int): Int = if (n == 0) 0 else if (n == 1) -10 else 40

      val cache = Array.ofDim[Int](m, n, introvertsCount + 1, extrovertsCount + 1, memSize)

      def getMaxHappiness(row: Int, col: Int, is: Int, es: Int, mem: Int): Int = {
        if (is == 0 && es == 0) 0
        else if (col == n) getMaxHappiness(row + 1, 0, is, es, mem)
        else if (row == m) 0
        else if (cache(row)(col)(is)(es)(mem) != 0) cache(row)(col)(is)(es)(mem)
        else {
          val nCol = col + 1
          val up = if (row > 0) getUp(mem) else 0
          val left = if (col > 0) getLeft(mem) else 0

          val r0 =
            getMaxHappiness(row, nCol, is, es, addToMem(mem, 0))
          val r1 = if (is == 0) 0 else {
            val localHappiness = 120 + intH(left) + intH(up)
            getMaxHappiness(row, nCol, is - 1, es, addToMem(mem, 1)) + localHappiness
          }
          val r2 = if (es == 0) 0 else {
            val localHappiness = 40 + extH(left) + extH(up)
            getMaxHappiness(row, nCol, is, es - 1, addToMem(mem, 2)) + localHappiness
          }
          val res = r0 max r1 max r2
          cache(row)(col)(is)(es)(mem) = res
          res
        }
      }

      getMaxHappiness(0, 0, introvertsCount, extrovertsCount, 0)
    }
  }

  import Solution.getMaxGridHappiness

  "Example 1: (m = 2, n = 3, introvertsCount = 1, extrovertsCount = 2) -> 240" in {
    getMaxGridHappiness(2, 3, 1, 2) shouldBe 240
    //Explanation: Assume the grid is 1-indexed with coordinates (row, column).
    //We can put the introvert in cell (1,1) and put the extroverts in cells (1,3) and (2,3).
    //- Introvert at (1,1) happiness: 120 (starting happiness) - (0 * 30) (0 neighbors) = 120
    //- Extrovert at (1,3) happiness: 40 (starting happiness) + (1 * 20) (1 neighbor) = 60
    //- Extrovert at (2,3) happiness: 40 (starting happiness) + (1 * 20) (1 neighbor) = 60
    //The grid happiness is 120 + 60 + 60 = 240.
  }
  "Example 2: (m = 3, n = 1, introvertsCount = 2, extrovertsCount = 1) -> 260" in {
    getMaxGridHappiness(3, 1, 2, 1) shouldBe 260
    //Explanation: Place the two introverts in (1,1) and (3,1) and the extrovert at (2,1).
    //- Introvert at (1,1) happiness: 120 (starting happiness) - (1 * 30) (1 neighbor) = 90
    //- Extrovert at (2,1) happiness: 40 (starting happiness) + (2 * 20) (2 neighbors) = 80
    //- Introvert at (3,1) happiness: 120 (starting happiness) - (1 * 30) (1 neighbor) = 90
    //The grid happiness is 90 + 80 + 90 = 260.
  }
  "Example 3: (m = 2, n = 2, introvertsCount = 4, extrovertsCount = 0) -> 240" in {
    getMaxGridHappiness(2, 2, 4, 0) shouldBe 240
  }

  "simple grids" should {
    "(m = 1, n = 1, introvertsCount = 1, extrovertsCount = 0) -> 120" in {
      getMaxGridHappiness(1, 1, 1, 0) shouldBe 120
    }
    "(m = 1, n = 1, introvertsCount = 0, extrovertsCount = 1) -> 40" in {
      getMaxGridHappiness(1, 1, 0, 1) shouldBe 40
    }
    "(m = 1, n = 1, introvertsCount = 1, extrovertsCount = 1) -> 120" in {
      getMaxGridHappiness(1, 1, 1, 1) shouldBe 120
    }

    "(m = 2, n = 1, introvertsCount = 1, extrovertsCount = 1) -> 150" in {
      getMaxGridHappiness(2, 1, 1, 1) shouldBe 150
    }
    "(m = 1, n = 2, introvertsCount = 1, extrovertsCount = 1) -> 150" in {
      getMaxGridHappiness(1, 2, 1, 1) shouldBe 150
    }
    "(m = 2, n = 1, introvertsCount = 2, extrovertsCount = 0) -> 180" in {
      getMaxGridHappiness(2, 1, 2, 0) shouldBe 180
    }
    "(m = 1, n = 2, introvertsCount = 2, extrovertsCount = 0) -> 180" in {
      getMaxGridHappiness(1, 2, 2, 0) shouldBe 180
    }
    "(m = 2, n = 1, introvertsCount = 0, extrovertsCount = 2) -> 120" in {
      getMaxGridHappiness(2, 1, 0, 2) shouldBe 120
    }
    "(m = 1, n = 2, introvertsCount = 0, extrovertsCount = 2) -> 120" in {
      getMaxGridHappiness(1, 2, 0, 2) shouldBe 120
    }
  }
  "(m = 5, n = 5, introvertsCount = 6, extrovertsCount = 6) -> 1240" in {
    getMaxGridHappiness(5, 5, 6, 6) shouldBe 1240
  }
}

File: rust/src/contest/w215_4.rs
#![allow(dead_code)]

struct Solution;
struct Solution1;

impl Solution {
    fn get_max_grid_happiness(n: i32, m: i32, introverts_count: i32, extroverts_count: i32) -> i32 {
        let mut res = 0;
        let m = m as usize;
        let n = n as usize;
        let mut grid: Vec<Vec<i32>> = vec![vec![0; m]; n];
        let mut cur = 0;
        Self::dfs(
            0,
            introverts_count,
            extroverts_count,
            &mut grid,
            &mut cur,
            &mut res,
            n,
            m,
        );
        res
    }
    fn dfs(
        start: usize,
        a: i32,
        b: i32,
        grid: &mut Vec<Vec<i32>>,
        cur: &mut i32,
        max: &mut i32,
        n: usize,
        m: usize,
    ) {
        if a == 0 && b == 0 || start == n * m {
            *max = (*max).max(*cur);
            return;
        }
        if *cur + a * 120 + b * 120 <= *max { return; }
        let i = start / m;
        let j = start % m;
        if a > 0 {
            grid[i][j] = 1;
            let mut base = 120;
            if i > 0 && grid[i - 1][j] == 1 { base -= 60; }
            if i > 0 && grid[i - 1][j] == 2 { base -= 10; }
            if j > 0 && grid[i][j - 1] == 1 { base -= 60; }
            if j > 0 && grid[i][j - 1] == 2 { base -= 10; }
            *cur += base;
            Self::dfs(start + 1, a - 1, b, grid, cur, max, n, m);
            *cur -= base;
            grid[i][j] = 0;
        }
        if b > 0 {
            grid[i][j] = 2;
            let mut base = 40;
            if i > 0 && grid[i - 1][j] == 1 { base -= 10; }
            if i > 0 && grid[i - 1][j] == 2 { base += 40; }
            if j > 0 && grid[i][j - 1] == 1 { base -= 10; }
            if j > 0 && grid[i][j - 1] == 2 { base += 40; }
            *cur += base;
            Self::dfs(start + 1, a, b - 1, grid, cur, max, n, m);
            *cur -= base;
            grid[i][j] = 0;
        }
        if (i > 0 && grid[i - 1][j] != 0) || (j > 0 && grid[i][j - 1] != 0) {
            Self::dfs(start + 1, a, b, grid, cur, max, n, m);
        }
    }
}

impl Solution1 {
    pub fn get_max_grid_happiness(m: i32, n: i32, introverts_count: i32, extroverts_count: i32) -> i32 {
        struct Helpers {
            m: usize,
            n: usize,
            mem_size: usize,
            cache: Vec<Vec<Vec<Vec<Vec<i32>>>>>,
        }

        fn i_hap(n: usize) -> i32 { if n == 0 { 0 } else if n == 1 { -60 } else { -10 } }
        fn e_hap(n: usize) -> i32 { if n == 0 { 0 } else if n == 1 { -10 } else { 40 } }

        impl Helpers {
            fn new(m: usize, n: usize, ic: usize, ec: usize) -> Helpers {
                let mem_size = 3usize.pow(n as u32);
                let cache = vec![vec![vec![vec![vec![-1; mem_size]; ec + 1]; ic + 1]; n]; m];
                Helpers { m, n, mem_size, cache }
            }

            fn add_to_mem(&self, mem: usize, v: usize) -> usize { (mem * 3 + v) % self.mem_size }
            fn get_left(&self, mem: usize) -> usize { mem % 3 }
            fn get_up(&self, mem: usize) -> usize { mem / (self.mem_size / 3) }

            fn find_max(&mut self, row: usize, col: usize, ic: usize, ec: usize, mem: usize) -> i32 {
                if ic == 0 && ec == 0 {
                    0
                } else if row == self.m {
                    0
                } else if col == self.n {
                    self.find_max(row + 1, 0, ic, ec, mem)
                } else if self.cache[row][col][ic][ec][mem] != -1 {
                    self.cache[row][col][ic][ec][mem]
                } else {
                    let n_col = col + 1;
                    let up = if row == 0 { 0 } else { self.get_up(mem) };
                    let left = if col == 0 { 0 } else { self.get_left(mem) };

                    let hap0 = self.find_max(row, n_col, ic, ec, self.add_to_mem(mem, 0));
                    let hap1 = if ic == 0 { 0 } else {
                        let local_hap = 120 + i_hap(up) + i_hap(left);
                        let hap = self.find_max(row, n_col, ic - 1, ec, self.add_to_mem(mem, 1));
                        hap + local_hap
                    };
                    let hap2 = if ec == 0 { 0 } else {
                        let local_hap = 40 + e_hap(up) + e_hap(left);
                        let hap = self.find_max(row, n_col, ic, ec - 1, self.add_to_mem(mem, 2));
                        hap + local_hap
                    };

                    let hap = hap0.max(hap1).max(hap2);
                    self.cache[row as usize][col as usize][ic as usize][ec as usize][mem as usize] = hap;
                    hap
                }
            }
        }

        let mut h = Helpers::new(m as usize, n as usize, introverts_count as usize, extroverts_count as usize);

        h.find_max(0, 0, introverts_count as usize, extroverts_count as usize, 0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_2_3_1_2() {
        assert_eq!(Solution::get_max_grid_happiness(2, 3, 1, 2), 240);
    }

    #[test]
    fn test_3_1_2_1() {
        assert_eq!(Solution::get_max_grid_happiness(3, 1, 2, 1), 260);
    }

    #[test]
    fn test_2_2_4_0() {
        assert_eq!(Solution::get_max_grid_happiness(2, 2, 4, 0), 240);
    }

    #[test]
    fn test_5_5_6_6() {
        assert_eq!(Solution::get_max_grid_happiness(5, 5, 6, 6), 1240);
    }
}